server/src/repositories/Equipment/ReservationRepository.ts

```diff
@@ -54,32 +64,129 @@ export class ReservationRepository implements IReservationRepository {
+      // get all modules needed for the equipment
+      const modules = (
+        await this.queryBuilder("ModulesForEquipment")
+        .select("trainingModuleId")
+        .where("equipmentId", reservation.equipment)
+      );
+      let passed = true;
+      // get last submission from maker for every module
+      modules.forEach(async (module: any) => {
+        const eligibility = await this.queryBuilder("ModuleSubmissions")
+          .select("passed")
+          .where("moduleID", module.id)
+          .where("makerID", reservation.maker)
+          .orderBy("submissionDate", "desc")
+          .limit(1)
+          .first();
+        if (!eligibility) {
+          passed = false;
+        }
+      });
+      // in progress
+      // const startTime = reservation.startTime;
+      // const endTime
+      // const free = await this.queryBuilder("Reservations")
+      //   .select("*")
+      //   .whereBetween("startTime", [reservation.startTime, reservation.endTime])
+      //   .whereBetween("endTime", [reservation.startTime, reservation.endTime])
+      // if maker has passed all modules 
+      if (passed) {
+        const newId = (
+          await this.queryBuilder("Reservations").insert(
+            {
+              creator: reservation.creator,
+              labbie: reservation.labbie,
+              maker: reservation.maker,
+              equipment: reservation.equipment,
+              startTime: reservation.startTime,
+              endTime: reservation.endTime,
+              startingMakerComment: reservation.startingMakerComment
+            },
+            "id"
+          )
+        )[0];
+        return singleReservationToDomain(this.getReservationById(newId));
+      } else {
+        // idk if this is the right thing to do
+        return null;
+      }
     }
 
-    public async archiveReservation(id: number): Promise<void> {
-        await knex("Reservations").where({ id: id}).update({archived: true})
-    }
+  public async assignLabbieToReservation(resId: number, labbieId: number): Promise<Reservation | null> {
+    await this.queryBuilder("Reservations")
+      .where("id", resId)
+      .update({labbie: labbieId});
+    return singleReservationToDomain(this.getReservationById(resId));
+  }
+
+  public async removeLabbieFromReservation(resId: number): Promise<Reservation | null> {
+    await this.queryBuilder("Reservations")
+      .where("id", resId)
+      .update({labbie: null});
+    return singleReservationToDomain(this.getReservationById(resId));
+  }
+
+  public async addComment(resId: number, authorId: number, commentText: string): 
+  Promise<string | null> {
+    const newId = (
+      await this.queryBuilder("ReservationEvents").insert(
+        {
+          eventType: "COMMENT",
+          reservationId: resId,
+          user: authorId,
+          payload: commentText
+        },
+        "id"
+      )
+    )[0];
+    
+    // return last added comment
+    return knex("Reservations")
+    .join(
```

Comment:
Do we need this? Can't we just return the comment ID instead?

Reply:
We can, just wanted to see the comment itself lol